### ISSUE TITLE
Transitive compatibility rules may give false positives

### ISSUE THREAD

**Issue by jsenko** [Apicurio/apicurio-registry#342]
Transitive compatibility rules must check the whole sequence of schemas (artifact versions) each time it is executed. We should not be using singleton list here in that case:
https://github.com/Apicurio/apicurio-registry/blob/1.1.x/app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityRuleExecutor.java#L45

**Comment by carlesarnal** [comment]
@jsenko do you have any particular opinion on how to better fix this?

cc @EricWittmann

**Comment by EricWittmann** [comment]
I'm just concerned about the idea of querying the DB for *all* previous versions.  Such a list is unbounded, and that's a problem.  We should consider a streaming approach to providing the list of previous versions to the rules.  Perhaps a custom List that dynamically loads each version.  That's my main thought.

**Comment by jsenko** [comment]
That's a good point, agreed. I don't see any way of avoiding the loading of all of the versions. Two optimizations come to mind - loading in chunks, which could be faster for new artifacts, using only one query - and caching the content for a short period, which would however only help when multiple versions are added in a short time. Not sure if they are worth using though, unless we have concrete data on how often they would apply.

**Comment by jsenko** [comment]
The caching should be probably done in a more holistic way with https://github.com/Apicurio/apicurio-tasks/issues/1 in mind.

**Comment by EricWittmann** [comment]
I'm imagining something like a custom `List` implementation, that can be returned by the storage layer, which loads the items in the list on demand.  So basically it's a virtual list backed by the storage itself.

**Comment by EricWittmann** [comment]
I don't think this should be done as part of https://github.com/Apicurio/apicurio-tasks/issues/1 because that task is about externalized caching of content to make the application more robust (e.g. by having something like http://www.squid-cache.org/ sitting in front of the app).
